File: core/src/ru/myitschool/dcrawler/skills/Immobilize.java
package ru.myitschool.dcrawler.skills;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.effects.DisarmedEffect;
import ru.myitschool.dcrawler.effects.ImmobilizedEffect;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.skills.action.Action;
import ru.myitschool.dcrawler.skills.action.ComplexEffectAction;
import ru.myitschool.dcrawler.skills.action.SimpleEffectAction;
import ru.myitschool.dcrawler.skills.check.EntityPlayCheck;
import ru.myitschool.dcrawler.skills.patterns.EnemyTargetPattern;
import ru.myitschool.dcrawler.skills.patterns.TargetPattern;
import ru.myitschool.dcrawler.skills.targeting.EnemyDisplayer;
import ru.myitschool.dcrawler.skills.targeting.TargetDisplayer;

/**
 * Created by Voyager on 02.12.2017.
 */
public class Immobilize extends Skill {

    int normalTurns = 1;
    int criticalTurns = 2;

    public Immobilize(Entity doer) {
        super(doer);
        PlayContainer container = addPlayContainer();
        container.getEnemyPlay().setPlayCheck(new EntityPlayCheck(this));
        container.getEnemyPlay().addAction(new ComplexEffectAction(this) {
            int turns = 0;

            @Override
            protected void beforeEffect(Target target, int damage, FloatingDamageMark mark) {

            }

            @Override
            public void successEffect(Target target, int damage, FloatingDamageMark mark) {
                turns = normalTurns;
            }

            @Override
            protected void critSuccessEffect(Target target, int damage, FloatingDamageMark mark) {
                turns = criticalTurns;
            }

            @Override
            protected void failEffect(Target target, int damage, FloatingDamageMark mark) {

            }

            @Override
            protected void critFailEffect(Target target, int damage, FloatingDamageMark mark) {

            }

            @Override
            protected void afterEffect(Target target, int damage, FloatingDamageMark mark) {
                if (turns > 0) {
                    target.getEntity().addEffect(new ImmobilizedEffect(target.getEntity(), turns));
                    target.getEntity().addEffect(new DisarmedEffect(target.getEntity(), turns));
                }
                if (turns == normalTurns) {
                    mark.addText("Success");
                } else if (turns == criticalTurns) {
                    mark.addText("Critical success");
                }
            }
        });
    }

    @Override
    public void maintainDisplayers(Array<TargetDisplayer> displayers) {
        displayers.add(new EnemyDisplayer());
    }

    @Override
    public TargetPattern getPattern() {
        return new EnemyTargetPattern(this);
    }

    @Override
    public Texture getIcon() {
        return new Texture("immobilization.png");
    }

    @Override
    public String getName() {
        return "Immobilize";
    }

    @Override
    public String getDescription() {
        return "You immobilizes your target for " + normalTurns + "(" + criticalTurns + ")";
    }

    @Override
    public int getSkillAccuracyBonus() {
        return 0;
    }

    @Override
    public int getRange() {
        return 1;
    }

    @Override
    public int getDistanceMin() {
        return 1;
    }

    @Override
    public int getDistanceMax() {
        return 8;
    }

    @Override
    public int getTargetCountMax() {
        return 1;
    }

    @Override
    public int getCooldownMax() {
        return 2;
    }

    @Override
    public int getType() {
        return SKILL_TYPE_COOLDOWN_DICE;
    }

    @Override
    public int getTargetType() {
        return SKILL_TARGET_TYPE_ENEMY;
    }

    @Override
    public boolean isCheckAllTargets() {
        return false;
    }

    @Override
    public boolean isMarkEverything() {
        return false;
    }

    @Override
    public boolean isMark() {
        return true;
    }

    @Override
    public boolean isObstruct() {
        return false;
    }

    @Override
    public boolean isWallTargets() {
        return false;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/Immobilize.kt
package com.wolg_vlad.dcrawler.entities.skills

import com.badlogic.gdx.graphics.Texture
import com.wolg_vlad.dcrawler.effects.DisarmedEffect
import com.wolg_vlad.dcrawler.effects.ImmobilizedEffect
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.skills.action.ComplexEffectAction
import com.wolg_vlad.dcrawler.entities.skills.check.EntityPlayCheck
import com.wolg_vlad.dcrawler.entities.skills.patterns.EnemyTargetPattern
import com.wolg_vlad.dcrawler.entities.skills.patterns.TargetPattern
import com.wolg_vlad.dcrawler.entities.skills.targeting.EnemyDisplayer
import com.wolg_vlad.dcrawler.entities.skills.targeting.TargetDisplayer

/**
 * Created by Voyager on 02.12.2017.
 */
class Immobilize(doer: Entity) : Skill(doer) {
    var normalTurns = 1
    var criticalTurns = 2

    override val displayers: Array<TargetDisplayer> = arrayOf(EnemyDisplayer())
    override val pattern: TargetPattern = EnemyTargetPattern(this)
    override val icon: Texture = Texture("immobilization.png")
    override val name: String = "Immobilize"
    override val description: String = "You immobilizes your target for $normalTurns($criticalTurns)"
    override val skillAccuracyBonus: Int = 0
    override val range: Int = 1
    override val distanceMin: Int = 1
    override val distanceMax: Int = 8
    override val targetCountMax: Int = 1
    override val cooldownMax: Int = 2
    override val type: Int = SKILL_TYPE_COOLDOWN_DICE
    override val targetType: Int = SKILL_TARGET_TYPE_ENEMY
    override val isCheckAllTargets: Boolean = false
    override val isMarkEverything: Boolean = false
    override val isMark: Boolean = true
    override val isObstruct: Boolean = false
    override val isWallTargets: Boolean = false

    init {
        val container = addPlayContainer()
        container.enemyPlay.setPlayCheck(EntityPlayCheck(this))
        container.enemyPlay.addAction(object : ComplexEffectAction(this) {
            var turns = 0
            override fun beforeEffect(target: Target?, damage: Int, mark: FloatingDamageMark?) {}
            public override fun successEffect(target: Target, damage: Int, mark: FloatingDamageMark) {
                turns = normalTurns
            }

            override fun critSuccessEffect(target: Target?, damage: Int, mark: FloatingDamageMark?) {
                turns = criticalTurns
            }

            override fun failEffect(target: Target?, damage: Int, mark: FloatingDamageMark?) {}
            override fun critFailEffect(target: Target?, damage: Int, mark: FloatingDamageMark?) {}
            override fun afterEffect(target: Target, damage: Int, mark: FloatingDamageMark) {
                if (turns > 0) {
                    target.entity!!.addEffect(ImmobilizedEffect(target.entity, turns))
                    target.entity!!.addEffect(DisarmedEffect(target.entity, turns))
                }
                if (turns == normalTurns) {
                    mark.addText("Success")
                } else if (turns == criticalTurns) {
                    mark.addText("Critical success")
                }
            }
        })
    }
}